crates/ecs/src/lib.rs

```diff
@@ -457,11 +461,17 @@ pub enum ArchetypeError {
 /// Whether a archetype operation succeeded.
 pub type ArchetypeResult<T, E = ArchetypeError> = Result<T, E>;
 
+/// A hashmap where the keys are not hashed
+type NoHashHashMap<T, S> = HashMap<T, S, BuildHasherDefault<NoHashHasher<T>>>;
+
+/// A hashset where the keys are not hashed
+type NoHashHashSet<T> = HashSet<T, BuildHasherDefault<NoHashHasher<T>>>;
+
 /// Stores components associated with entity ids.
 #[derive(Debug, Default)]
 struct Archetype {
-    component_typeid_to_component_vec: HashMap<TypeId, Box<dyn ComponentVec>>,
-    entity_to_component_index: HashMap<Entity, ComponentIndex>,
+    component_typeid_to_component_vec: FnvHashMap<TypeId, Box<dyn ComponentVec>>,
+    entity_to_component_index: NoHashHashMap<Entity, ComponentIndex>,
     entity_order: Vec<Entity>,
 }
```

Comment:
The function `store_entity` can be replaced with this to to get rid of the unnecessary call to `entity_to_component_index.contains_key(&entity)`.
```rust
fn store_entity(&mut self, entity: Entity) -> ArchetypeResult<()> {
    let entity_index = self.entity_to_component_index.len();

    match self.entity_to_component_index.insert(entity, entity_index) {
        None => {
            self.entity_order.push(entity);
            Ok(())
        }
        Some(_) => {
            Err(ArchetypeError::EntityAlreadyExists(entity))
        }
    }
}
```